### PR TITLE
NickAkhmetov/CAT-844 Migrate away from using `descendants` and `ancestors` in favor of `id` equivalents

### PR DESCRIPTION
**Pull request by NickAkhmetov**
## Summary

This PR removes the portal UI's dependencies on the direct ancestor/descendant information available in the entity API in favor of using the corresponding `ids`.

### Progress
- `ancestors`
	- [x] Dev search used `ancestors` to determine raw/processed status -> revised to use `processing` field
	- [x] Prov table used `ancestors` to construct the table -> revised to use `ancestor_ids`, corresponding revisions made to error display for donors
	- [x] Ancestors' `rui_location` field is used for `details_rui_json` on back-end and to determine whether a dataset is spatial in the dev search
	- [x] Ancestors not being datasets is currently a condition for the unified prov view lookup when looking up the top level dataset -> dropped the condition in favor of using `processing: raw`, there should be 0..1 raw datasets in any given dataset's ancestors anyway.
- `descendants`
  - Not in use.
- `immediate_ancestors`
  - Mentioned in an artillery config, but otherwise not present.
- `immediate_descendants`
  - ~~`useImmediateDescendantProv` is used for the `view derived entities` button~~ This actually uses the entity API endpoint; this can be improved to use the search API equivalent once the upstream change is made to the search API, but is not currently a blocker.

## Design Documentation/Original Tickets

https://hms-dbmi.atlassian.net/browse/CAT-844

## Testing

Manually verified, confirmed functionality is analogous to current production behaviors.

## Screenshots/Video

Not very informative since this matches existing functionality, but for verification:

image

<details><summary>Dev search filters</summary>
<p>

image

image

image

</p>
</details> 

## Checklist

- [x] Code follows the project's coding standards
  - [x] Lint checks pass locally
  - [x] New `CHANGELOG-your-feature-name-here.md` is present in the root directory, describing the change(s) in full sentences.
- [x] Unit tests covering the new feature have been added
- [x] All existing tests pass
- [x] Any relevant documentation in JIRA/Confluence has been updated to reflect the new feature
- [x] Any new functionalities have appropriate analytics functionalities added

### CODE REVIEW

File: context/app/routes_browse.py [hunk 1]
```diff
@@ -139,18 +139,12 @@ def details_rui_json(type, uuid):
         abort(404)
     client = get_client()
     entity = client.get_entity(uuid)
-    # For Samples...
+    # For samples and datasets, the nearest RUI location is indexed with the entity itself.
+    # https://github.com/hubmapconsortium/search-api/pull/860
     if 'rui_location' in entity:
         return json.loads(entity['rui_location'])
```
Comment (john-conroy) [inline context/app/routes_browse.py:145]: Datasets don't include an `rui_location` unless you've made some corresponding change in the `search-api`.

File: context/app/static/js/components/entity-tile/EntityTile/EntityTile.tsx [hunk 2]
```diff
@@ -73,56 +73,25 @@ function ErrorTile({ entity_type, id }: Pick<EntityTileProps, 'id' | 'entity_typ
...
       variant="outlined"
     >
-      <Stack direction="row" gap={1} color="error.main" alignItems="center">
-        <StatusIcon status="error" sx={{ fontSize: '1.5rem' }} />
-        <Typography variant="subtitle1">{id}</Typography>
-        <IconButton onClick={() => copy(id)}>
-          <ContentCopyIcon color="info" />
-        </IconButton>
+      <Stack direction="row" gap={1}>
+        <StatusIcon status="error" sx={{ alignSelf: 'start', fontSize: '1.3rem' }} />
```
Comment (john-conroy) [inline context/app/static/js/components/entity-tile/EntityTile/EntityTile.tsx:77]: Maybe `1.25rem`, I believe we use that more commonly throughout. I can't recall using `1.3rem` before, but maybe you have?

File: context/app/static/js/components/types.ts [hunk 3]
```diff
@@ -38,7 +38,9 @@ export interface Entity {
   contacts: ContactAPIResponse[];
   contributors: ContributorAPIResponse[];
   created_timestamp: number;
+  /** @deprecated Use `ancestor_ids` and `useEntitiesData` instead */
   ancestors: Entity[];
+  ancestor_ids: string[];
   // eslint-disable-next-line no-use-before-define -- Donor is defined later in the file and extends Entity
   donor: Donor;
   descendant_counts: { entity_type: Record<string, number> };
```
Comment (john-conroy) [inline context/app/static/js/components/types.ts:46]: Should we add like comments for `descendants` and `descendant_counts` or do we plan to keep those for now? I don't think we have the immediate variants in the types.

File: context/app/static/js/hooks/useEntityData.ts [hunk 4]
```diff
@@ -2,12 +2,30 @@ import { useMemo } from 'react';
...
 
   return [searchHits[0]?._source, isLoading];
 }
 
+export function useEntitiesData(uuids: string[], source?: string[]): [Entity[], boolean] {
+  const query = useEntityQuery(uuids, source);
+
+  const { searchHits, isLoading } = useSearchHits<Entity>(query);
+
+  return [searchHits.map((hit) => hit._source), isLoading];
```
Comment (john-conroy) [inline context/app/static/js/hooks/useEntityData.ts:28]: Slightly prefer an obj so we can also include the other returned values from `useSearchHits`.

File: context/app/static/js/pages/search/DevSearch.jsx [hunk 5]
```diff
@@ -88,10 +88,8 @@ function DevSearch() {
         checkboxFilter('no_metadata', 'No metadata?', BoolMustNot(ExistsQuery('metadata.metadata'))),
         checkboxFilter('has_files', 'Has files?', ExistsQuery('files')),
         checkboxFilter('no_files', 'No files?', BoolMustNot(ExistsQuery('files'))),
-        checkboxFilter('has_rui_sample', 'Spatial Sample?', ExistsQuery('rui_location')),
-        checkboxFilter('no_rui_sample', 'Not Spatial Sample?', BoolMustNot(ExistsQuery('rui_location'))),
-        checkboxFilter('has_rui_dataset', 'Spatial Dataset?', ExistsQuery('ancestors.rui_location')),
-        checkboxFilter('no_rui_dataset', 'Not Spatial Dataset?', BoolMustNot(ExistsQuery('ancestors.rui_location'))),
+        checkboxFilter('is_spatial', 'Spatial?', ExistsQuery('rui_location')),
+        checkboxFilter('not_spatial', 'Not Spatial?', BoolMustNot(ExistsQuery('rui_location'))),
```
Comment (john-conroy) [inline context/app/static/js/pages/search/DevSearch.jsx:92]: Like above, I think this won't work for datasets unless there was some change I'm not aware of?

**Review by john-conroy — COMMENTED** [review]
Do we plan to keep `entity_type` in the descendants for `descendants_counts`?

File: context/app/routes_browse.py (hunk 1 above)
Comment (NickAkhmetov) [reply context/app/routes_browse.py:145]: I did make this change: https://github.com/hubmapconsortium/search-api/blob/dev-integrate/src/hubmap_translation/addl_index_transformations/portal/translate.py#L307

File: context/app/static/js/components/types.ts (hunk 3 above)
Comment (NickAkhmetov) [reply context/app/static/js/components/types.ts:46]: I think we do plan to keep `descendant_counts`,for the time being, but good call on adding this comment for `descendants`!

File: context/app/static/js/hooks/useEntityData.ts (hunk 4 above)
Comment (NickAkhmetov) [reply context/app/static/js/hooks/useEntityData.ts:28]: >other returned values

From what I can tell, `useSearchHits` only returns the `searchHits` and `isLoading` values, so I stuck with the same format that `useEntityData` uses above. Are you interested in us keeping the top-level `hit` data?

File: context/app/static/js/pages/search/DevSearch.jsx (hunk 5 above)
Comment (NickAkhmetov) [reply context/app/static/js/pages/search/DevSearch.jsx:92]: This is now supported with the change I linked earlier.

**Review by john-conroy — APPROVED** [review]

**Pull request merged**